src/lib/utils/bitvector.h

```diff
@@ -0,0 +1,1327 @@
+
+      void resize(size_type bits) {
+         const auto new_number_of_blocks = ceil_toblocks(bits);
+         if(new_number_of_blocks != m_blocks.size()) {
+            m_blocks.resize(new_number_of_blocks);
+         }
+
+         m_bits = bits;
+         zero_unused_bits();
+      }
+
+      void push_back(bool bit) {
+         const auto i = size();
+         resize(i + 1);
+         ref(i) = bit;
+      }
+
+      void pop_back() {
+         if(!empty()) {
+            resize(size() - 1);
+         }
+      }
+
+      /// @}
+
+      /// @name Bitwise and Global Accessors and Modifiers
+      /// @{
+
+      auto at(size_type pos) {
+         check_offset(pos);
+         return ref(pos);
+      }
+
+      // TODO C++23: deducing this
+      auto at(size_type pos) const {
+         check_offset(pos);
+         return ref(pos);
+      }
+
+      auto front() { return ref(0); }
+
+      // TODO C++23: deducing this
+      auto front() const { return ref(0); }
+
+      auto back() { return ref(size() - 1); }
+
+      // TODO C++23: deducing this
+      auto back() const { return ref(size() - 1); }
+
+      /**
+       * Sets the bit at position @p pos.
+       * @throws Botan::Invalid_Argument if @p pos is out of range
+       */
+      bitvector_base& set(size_type pos) {
+         check_offset(pos);
+         ref(pos).set();
+         return *this;
+      }
+
+      /**
+       * Sets all currently allocated bits.
+       */
+      bitvector_base& set() {
+         full_range_operation(
+            [](std::unsigned_integral auto block) -> decltype(block) { return static_cast<decltype(block)>(~0); },
+            *this);
+         zero_unused_bits();
+         return *this;
+      }
+
+      /**
+       * Unsets the bit at position @p pos.
+       * @throws Botan::Invalid_Argument if @p pos is out of range
+       */
+      bitvector_base& unset(size_type pos) {
+         check_offset(pos);
+         ref(pos).unset();
+         return *this;
+      }
+
+      /**
+       * Unsets all currently allocated bits.
+       */
+      bitvector_base& unset() {
+         full_range_operation(
+            [](std::unsigned_integral auto block) -> decltype(block) { return static_cast<decltype(block)>(0); },
+            *this);
+         return *this;
+      }
+
+      /**
+       * Flips the bit at position @p pos.
+       * @throws Botan::Invalid_Argument if @p pos is out of range
+       */
+      bitvector_base& flip(size_type pos) {
+         check_offset(pos);
+         ref(pos).flip();
+         return *this;
+      }
+
+      /**
+       * Flips all currently allocated bits.
+       */
+      bitvector_base& flip() {
+         full_range_operation([](std::unsigned_integral auto block) -> decltype(block) { return ~block; }, *this);
+         zero_unused_bits();
+         return *this;
+      }
+
+      /**
+       * @returns true iff no bit is set
+       */
+      bool none() const {
```

Comment:
This convention of `foo` plus `ct_foo` seems very easy to make a mistake and call `foo` in the wrong context and introduce a side channel.

In newer code (eg in pcurves) I've moved to `foo` is the constant time implementation, and if a variable time foo is required this is explicit as eg `vartime_foo` or `foo_vartime`. The explicit `vartime` jumps out at the reader and it's much harder to accidentally be variable time when not intended.

Reply:
I now applied this convention to `none()`, `all()`, `any()` and `equals()` and removed the `ct_` prefix from `hamming_weight()` and `has_odd_hamming_weight()` (without providing a ct-alternative). All other (non re-allocating) methods should be CT (review appreciated for this statement).